File: src/main/java/com/anexsys/recruitment/challenge/Player.java
package com.anexsys.recruitment.challenge;

import java.util.Set;
import java.util.HashSet;

class Player {
    private final String id;
    private final Set<Card> currentCards;
    Player(String id) {
        this.id = id;
        this.currentCards = new HashSet<>();
    }
    Player receive(Card card) {
        this.currentCards.add(card);
        return this;
    }
    Player use(Card card) {
        boolean removed = this.currentCards.remove(card);
        if (!removed) {
            throw new IllegalArgumentException("Cannot use a card the Player does not have");
        }
        return this;
    }
    @Override
    public String toString() {
        return "player-" + this.id;
    }
}

File: CSharp/CSharp/Player.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp
{
    class Player
    {
        private readonly string id;
        private readonly ISet<Card> currentCards;

        public Player(string id)
        {
            this.id = id;
            currentCards = new HashSet<Card>();
        }

        public Player Receive(Card card)
        {
            currentCards.Add(card);
            return this;
        }

        public Player Use(Card card)
        {
            bool removed = currentCards.Remove(card);
            if (!removed)
            {
                throw new ArgumentOutOfRangeException("Cannot use a card the player does not have");
            }
            return this;
        }

        public override string ToString()
        {
            return $"player-{id}";
        }
    }
}
